L1Trigger/TrackFindingTracklet/test/AnalyzerTracklet.cc

```diff
@@ -0,0 +1,247 @@
+    }
+    // analyze tracklet products and associate found tracks with reconstrucable TrackingParticles
+    set<TPPtr> tpPtrs;
+    set<TPPtr> tpPtrsSelection;
+    int allMatched(0);
+    vector<vector<TTStubRef>> tracks;
+    tracks.reserve(ttTracks.size());
+    transform(ttTracks.begin(), ttTracks.end(), back_inserter(tracks), [](const TTTrack<Ref_Phase2TrackerDigi_>& ttTrack){ return ttTrack.getStubRefs(); });
+    if (useMCTruth_) {
+      int tmp(0);
+      associate(tracks, selection, tpPtrsSelection, tmp);
+      associate(tracks, reconstructable, tpPtrs, allMatched);
+    }
+    for (const TPPtr& tpPtr : tpPtrsSelection)
+      fill(tpPtr, hisEff_);
+    prof_->Fill(4, allMatched);
+    prof_->Fill(5, ttTracks.size());
+    prof_->Fill(6, tpPtrs.size());
+    prof_->Fill(7, tpPtrsSelection.size());
+    prof_->Fill(8, 0);
+    nEvents_++;
+  }
+
+  void AnalyzerTracklet::endJob() {
+    // effi
+    eff_->SetPassedHistogram(*hisEff_, "f");
+    eff_->SetTotalHistogram (*hisEffTotal_, "f");
+    // printout SF summary
+    const double totalTPs = prof_->GetBinContent(9);
+    const double numStubs = prof_->GetBinContent(1);
+    const double numTracks = prof_->GetBinContent(2);
+    const double totalTracks = prof_->GetBinContent(5);
+    const double numTracksMatched = prof_->GetBinContent(4);
+    const double numTPsAll = prof_->GetBinContent(6);
+    const double numTPsEff = prof_->GetBinContent(7);
+    const double errStubs = prof_->GetBinError(1);
+    const double errTracks = prof_->GetBinError(2);
+    const double fracFake = (totalTracks - numTracksMatched) / totalTracks;
+    const double fracDup = (numTracksMatched - numTPsAll) / totalTracks;
+    const double eff = numTPsEff / totalTPs;
+    const double errEff = sqrt(eff * (1. - eff) / totalTPs / nEvents_);
+    const vector<double> nums = {numStubs, numTracks};
+    const vector<double> errs = {errStubs, errTracks};
+    const int wNums = ceil(log10(*max_element(nums.begin(), nums.end()))) + 5;
+    const int wErrs = ceil(log10(*max_element(errs.begin(), errs.end()))) + 5;
+    log_ << "                      Tracklet  SUMMARY                      " << endl;
+    log_ << "number of stubs       per TFP = " << setw(wNums) << numStubs << " +- " << setw(wErrs) << errStubs << endl;
+    log_ << "number of tracks      per TFP = " << setw(wNums) << numTracks << " +- " << setw(wErrs) << errTracks << endl;
+    log_ << "          tracking efficiency = " << setw(wNums) << eff << " +- " << setw(wErrs) << errEff << endl;
+    log_ << "                    fake rate = " << setw(wNums) << fracFake << endl;
+    log_ << "               duplicate rate = " << setw(wNums) << fracDup << endl;
+    log_ << "=============================================================";
+    LogPrint("L1Trigger/TrackFindingTracklet") << log_.str();
+  }
+
+  //
+  void AnalyzerTracklet::associate(const vector<vector<TTStubRef>>& tracks, const StubAssociation* ass, set<TPPtr>& tps, int& sum) const {
```

Comment:
Add comment explaining associate(). e.g. "Gets all TPs associated too any of the tracks & number of tracks matching at least one TP". Also suggest renaming "sum" to "nMatchTrk" or similar.

Reply:
done